### ISSUE TITLE
[test-triage] chip_csr_aliasing chip_same_csr_outstanding chip_csr_bit_bash chip_sw_flash_ctrl_lc_rw_en RunThenComplete_M

### ISSUE THREAD

**Issue by johngt** [lowRISC/opentitan#17094]
### Hierarchy of regression failure

Chip Level

### Failure Description

```
Test chip_csr_aliasing has 3 failures.

0.chip_csr_aliasing.469089594
Line 920, in log /container/opentitan-public/scratch/os_regression/chip_earlgrey_asic-sim-vcs/0.chip_csr_aliasing/latest/run.log
```

```
  UVM_ERROR @ 34864.192652 us: (sha3pad.sv:851) [ASSERT FAILED] RunThenComplete_M
  UVM_INFO @ 34864.192652 us: (uvm_report_catcher.svh:705) [UVM/REPORT/CATCHER]
  --- UVM Report catcher Summary ---
```

### Steps to Reproduce

- GitHub Revision: [630616151](https://github.com/lowrisc/opentitan/tree/630616151de935942f78bfbb8c9d1848637ddb39)
- dvsim invocation command to reproduce the failure, inclusive of build and run seeds:
  ./util/dvsim/dvsim.py hw/top_earlgrey/dv/chip_sim_cfg.hjson -i chip_csr_aliasing --build-seed 20193871 --fixed-seed 469089594 --waves fsdb

### Tests with similar or related failures

- [ ] chip_csr_aliasing
- [ ] chip_same_csr_outstanding
- [ ] chip_csr_bit_bash
- [ ] chip_sw_flash_ctrl_lc_rw_en
- [ ] chip_sw_sram_ctrl_scrambled_access
- [ ] chip_sw_sram_ctrl_scrambled_access_jitter_en
- [ ] chip_sw_sram_ctrl_scrambled_access_jitter_en_reduced_freq
- [ ] chip_sw_exit_test_unlocked_bootstrap
- [ ] chip_prim_tl_access
- [ ] chip_tl_errors
- [ ] chip_csr_mem_rw_with_rand_reset

**Comment by johngt** [comment]
@vogelpi - tagging for assisting @alees24

**Comment by johngt** [comment]
Same error is also reported in chip_same_csr_outstanding test.

**Comment by johngt** [comment]
Looks like the implicated code is not in CSR specifically but rather sha3pad.sv.
sha3pad.sv:851 referenced in 121 failures.

**Comment by alees24** [comment]
chip_sw_sram_ctrl_scrambled_access is a good test to run because it reproduces this failure quickly. What I see from the run log is:

> TEST PASSED CHECKS
> ....
> $finish called from file "/nas/lowrisc/tools/synopsys/vcs/T-2022.06-SP2/etc/uvm-1.2/base/uvm_root.svh", line 527.
> $finish at simulation time 4015.828482 us
> "../src/lowrisc_ip_sha3_0.1/rtl/sha3pad.sv", 851: tb.dut.top_earlgrey.u_kmac.u_sha3.u_pad.RunThenComplete_M: started at 4014851924ps failed at 4015828482ps

The test code reports a pass and the clock to kmac is removed, reset is asserted and the clock is reinstated at a lower frequency (half?). Shortly after reset removal, kmac comes out of reset and responds to an 'app_i' request from u_rom_ctrl.

That's why kmac/sha3pad has started up and keccak_run_o is asserted circa 50 cycles (eyeballing) before the test finishes. There's no corresponding assertion of keccak_complete_i; earlier activations seem to take ca. 75 cycles to complete.

At this point, bearing in mind that I'm still very much feeling my way around, I'm hoping somebody else knows the intended behaviour. @vogelpi ?

**Comment by alees24** [comment]
I think this is beyond my present understanding of UVM and/or our DV environment to be able to unpick in a reasonable time frame. SW is signalling successful completion and the sw_status_if.sv reports successful completion at 4006us. The sha3pad assertion detects keccak_run_o at 4014.85us and then at 4015.83us the UVM report is generated twice
 'Quit count 0' reports a pass, the assertion fails, and the second report ('Quit count 1 of 1') shows the failed assertion...all at the same simulation time of 4015.83us.

I'm sure I'm missing something, but I don't see any evidence of a delay being introduced; can you explain why/how the 'strong' qualifier is supposed to work, and where I should look for the intended delay please; is the test software on the Ibex intended to delay signalling its status, or is there supposed to be something in our chip test/DV framework that waits when the first 'quit' occurs?

Re your last observation, the 'Share' parameter of the sha3pad module appears to be unchanged at '2' in the masked case, although that affect only the lower bound of the assertion [24*Share:$]

[Snippet from run_log.txt](https://github.com/lowRISC/opentitan/files/10435937/Snippet.from.run_log.txt)
![RunThenCompleteWaves](https://user-images.githubusercontent.com/117650394/212923813-dc05187b-f387-4189-accb-6c8c677785c5.png)

**Comment by engdoreis** [comment]
This assert failure is responsible for a significant amount of failing tests in the regression, so we should prioritise it.
@cindychip, as you have commits editing this assertion, would you be able to help us understand why it started failing now and what could be done?
@jadephilipoom and @johannheyszl could the recent crypto-related changes have had an impact?
The failures have been happening since around January 10th.

**Comment by jadephilipoom** [comment]
> @jadephilipoom and @johannheyszl could the recent crypto-related changes have had an impact? The failures have been happening since around January 10th.

Actually, it may be this: https://github.com/lowRISC/opentitan/pull/16909

That PR changed the way the KMAC dif executes `absorb`, which could have changed the timing of KMAC tests. I don't understand the context of this issue enough to know why that would cause the problem, but the timing seems to line up.

**Comment by cindychip** [comment]
> > @jadephilipoom and @johannheyszl could the recent crypto-related changes have had an impact? The failures have been happening since around January 10th.
> 
> Actually, it may be this: #16909
> 
> That PR changed the way the KMAC dif executes `absorb`, which could have changed the timing of KMAC tests. I don't understand the context of this issue enough to know why that would cause the problem, but the timing seems to line up.

Thanks for the reference. I will take a look!

**Comment by cindychip** [comment]
@jadephilipoom I dropped your commits and the error persists, so I do not think this failure relates to your commit :)
As far as I can tell, https://github.com/lowRISC/opentitan/commits/11ab36b4e7ad7a949a0a55bf42b161abe03d74b8
In Jan 10th, we committed something that causes the rom_ctrl to request two KMAC transactions, and the second transaction did not finish before the test exit. So the assertion failed.
I am confirming why we requested two rom_ctrl requests, and if it is reasonable, maybe a simple solution would be just to disable this assertion.

**Comment by cindychip** [comment]
I think this issue relates to VCS version.

I remembered on Jan 10th we moved to a newer Version of VCS. I believe it fixed an issue about assertions with `strong` type. In previous version - if a test exit with a `strong` assertion still running, it won't report an error. But in the latest version, it fixed the issue and will return an error.

I used the old VCS version to run the CSR tests, and did not see this assertion failure anymore.

As for the solution:
In chip level, we do not wait for all KMAC transactions to finish before exiting the test. If the main functionality of this test is checked, then we should be safe to exit.
So I think a better solution would be to move these assertions to block level test.
